frontend/apps/reader/modules/readerlink.lua

```diff
@@ -837,6 +880,17 @@ function ReaderLink:onGoBackLink(show_notification_if_empty)
     end
 end
 
+--- Goes to next location.
+function ReaderLink:onGoForwardLink()
+    local saved_location = table.remove(self.forward_location_stack)
+    if saved_location then
+        table.insert(self.location_stack, saved_location)
+        logger.dbg("GoForward: restoring:", saved_location)
+        self.ui:handleEvent(Event:new('RestoreBookLocation', saved_location))
```

Comment:
`saved_location` is the future forward location we are going to.
So, `table.insert(self.location_stack, saved_location)` seems wrong ? You should insert the current location we are leaving.

Reply:
I will start with going back (your other comment)
For example we have 3 previous locations on the stack, no forward locations, we pop one. If the current and popped are not the same we add current to forward stack. The popped location is added to the forward stack and we go to it. So now it is forward of us.

When we go forward we pop the last forward location off the stack, push it to previous location stack, and go to it.

Reply:
> we add current to forward stack. The popped location is added to the forward stack and we go to it. So now it is forward of us.

Still confused, sorry :/
When going back, we'll add 2 items to the foward_stack: the footnote page we were in, + the reading page we are going back to ? So the forward stack top elment is the current page ?

> When we go forward we pop the last forward location off the stack, push it to previous location stack, and go to it.

You're saying what I said: the previous location stack will be where we are going?
So, we don't add anywhere the current location ?!

Can you give some example with PDF pages (to keep it simpler than xpointer), the state of both stacks, when say:
1) I am on page 4, I follow a footnote link to page 12
2) then follow another link to page 17
3) then "Go back" and end up on page 12
4) then "Go back" and end up on page 4
5) then Go forward" and end up on page 12
6) then "Go back" and end up on page 4

Reply:
`location_stack = {}, forward_location_stack = {}`
I am on page 4, I follow a footnote link to page 12
`location_stack = { 4 }, forward_location_stack = {}`
then follow another link to page 17
`location_stack = { 4, 12 }, forward_location_stack = {}`
then "Go back" and end up on page 12
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
 then "Go back" and end up on page 4
`location_stack = {}, forward_location_stack = { 17, 12, 4 }`
then Go forward" and end up on page ~~12~~ **4**
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
then Go forward" and end up on page 12
`location_stack = { 4, 12 }, forward_location_stack = { 17 }`
then "Go back" and end up on page ~~4~~ **12**
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
then "Go back" and end up on page 4
`location_stack = {}, forward_location_stack = { 17, 12, 4 }`

Since we are always between locations (we store results, not actions) when you switch from forward<-->back, you repeat the first one going the other way. however if you move pages in the middle it make more sense:

`location_stack = {}, forward_location_stack = {}`
I am on page 4, I follow a footnote link to page 12
`location_stack = { 4 }, forward_location_stack = {}`
then follow another link to page 17
`location_stack = { 4, 12 }, forward_location_stack = {}`
then "Go back" and end up on page 12
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
 then "Go back" and end up on page 4
`location_stack = {}, forward_location_stack = { 17, 12, 4 }`
then Go forward" and end up on page ~~12~~ **4**
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
then Go forward" and end up on page 12
`location_stack = { 4, 12 }, forward_location_stack = { 17 }`
read to the next page (13) to finish the footnote.
then "Go back" and end up on page ~~4~~ **12**
`location_stack = { 4 }, forward_location_stack = { 17, 12 }`
then "Go back" and end up on page 4
`location_stack = {}, forward_location_stack = { 17, 12, 4 }`

If you want we can check if the `end of the stack == current_page and #location_stack >= 2` to jump 2 locations, solving the issue in the first case.

Reply:
I agree that the (page changing) action that does not make `addCurrentLocationToStack`, should modify the "forward" location.
Is it a "page turn" only?

Reply:
> In my mind, page 12 shouldn't be tracked:
if I am on page 4, then jump to page 12, then turn page to page 13, only 13 should be the forward location. So, going back and getting on page 4, only page 13 should be in the forward location - and when I "go forward", I reach page 13 directly.

Page 12 is not tracked, only if you first jumped to pg 17, and put pg 12 on the stack as in my earlier case, is it tracked

Reply:
OK, so the follow link & "Go back" workflow behaviour is not changed, if I never use "Go forward", right ?

It's just as soon as you use "Go forward" that more wanted or unwanted intermediate locations may be added to both stacks - and I guess the more of these the better? (as interested users get them, and non-interested users can just issue additional go back/forward to pass over them).
As my reactionnary mind is totally impermeable to progress and going forward :), I guess I'm in no position to contribute about how "go forward" and its stack behaviour should work :/

Reply:
If nobody else has remarks about your behaviour and we go with it, it would be good to have more comments in the code about this behaviour, or at least its tricky parts, with the kind of samples "if I go..." and the stacks states we used above, and "this allows... such use case".

Reply:
So, ask again after you've done that :)
(I won't put noise/directions into your genuine/naive appreciation of it all - but later, you can read the scenarios described above ("I am on page 4...").